File: gdb/domain/savings_account.py

from gdb.domain.abstract_account import AbstractAccount
from gdb.exceptions import MinimumBalanceViolationException

class SavingsAccount(AbstractAccount):
    def __init__(self, account_number, name, age, balance, status, pin, min_balance, interest_rate):
        super().__init__(account_number, name, age, balance, "SAVINGS", status, pin)
    
    def process_debit(self, amount: float) -> None:
        self.balance -= amount

File: src/com/gdb/domain/SavingsAccount.java
package com.gdb.domain;
import com.gdb.exceptions.*;

public class SavingsAccount extends AbstractAccount {
    private double minBalance;
    private double interestRate;

    public SavingsAccount(String accountNumber, String name, int age, double balance, String status, String pin, double minBalance, double interestRate) {
        super(accountNumber, name, age, balance, "SAVINGS", status, pin);
        this.minBalance = minBalance;
        this.interestRate = interestRate;
    }

    public void applyInterest() {
        this.balance += this.balance * (interestRate / 100.0);
    }

    @Override
    protected void processDebit(double amount) throws AccountException {
        if (this.balance - amount < minBalance) {
            throw new MinimumBalanceViolationException("Breaches minimum balance");
        }
        this.balance -= amount;
    }
}
